frontend/src/components/profile/info/BadgeHandler.ts

```diff
@@ -0,0 +1,252 @@
+            break;
+        case BADGE.DAILY_STREAK:
+            imageURL = `https://github.com/jalpp/DojoIcons/blob/main/milestones/DojoBadges/streak/clean/streak${level}paper-clean.png?raw=true`;
+            break;
+    }
+
+    return imageURL;
+}
+
+/**
+ * check if a user is eligible for the badge and given requirement
+ * @param user the current user
+ * @param reqid the requirement ID
+ * @param targetLevel the target level for user to earn the badge
+ * @returns if user is eligible
+ */
+function isEligibleForBadge(user: User, reqid: string, targetLevel: number): boolean {
+    const progress = user.progress[reqid];
+    if (!progress) {
+        return false;
+    }
+
+    let totalCount = 0;
+    for (const key in progress.counts) {
+        totalCount += progress.counts[key];
+    }
+
+    return totalCount >= targetLevel;
+}
+
+/**
+ * gets the level the user is elgible for
+ * @param user the current user
+ * @param reqid the requirement id
+ * @param levels the counting badge levels list
+ * @returns the current level badge user can get
+ */
+function isEligibleForLimit(user: User, reqid: string, levels: number[]): number {
+    let maxLevel: number = -1;
+    for (const level of levels) {
+        const elgible = isEligibleForBadge(user, reqid, level);
+
+        if (elgible) {
+            maxLevel = level;
+        } else {
+            return maxLevel;
+        }
+    }
+
+    return -1;
+}
+
+/**
+ * gets the badge message
+ * @param level the level for badge
+ * @param badge the badge type
+ * @returns the message
+ */
+function getBadgeMessage(level: number, badge: BADGE): string {
+    let msg: string;
+    switch (badge) {
+        case BADGE.POLGAR_MATE_ONE:
+            msg = polgarMateMilestones[level];
+            break;
+        case BADGE.POLGAR_MATE_TWO:
+            msg = polgarMateMilestones[level];
+            break;
+        case BADGE.POLGAR_MATE_THREE:
+            msg = polgarMateMilestones[level];
+            break;
+        case BADGE.ANNONTATE_GAMES:
+            msg = annotationMilestones[level];
+            break;
+        case BADGE.CLASSICAL_GAMES:
+            msg = classicalGameMilestones[level];
+            break;
+        case BADGE.DAILY_STREAK:
+            msg = 'Wowza! Your daily streak hit ' + level + '!';
+            break;
+    }
+
+    return msg;
+}
+
+/**
+ * gets the info and badge image for eligible badge
+ * @param user current user
+ * @param badge the badge type
+ * @returns info and image for the badge
+ */
+function isEligibleBadgeImage(user: User, badge: BADGE): string[] | undefined {
+    const level: number = isEligibleForLimit(user, badge, BADGE_LIMITS[badge]);
+    const info: string[] = [];
+    if (level == -1) {
+        return undefined;
+    }
+
+    info.push(getBadgeImage(level, badge));
+    info.push(BADGE_TITLE[badge] + ' ' + level);
+    info.push(getBadgeMessage(level, badge));
```

Comment:
Instead of arbitrarily pushing these in a random order, you should return an object like 

```
{
  image: getBadgeImage(level, badge),
  title: ...,
  message: ...,
}
```

The way you have it currently requires me to look at this function to know which order the strings are in. That order is easy to forget when calling the function and using the result elsewhere, which makes the rest of the code more error prone. If you return an object, then you will always know which is which.